### PR TITLE
feat: Support downloadable postman collection for API deployments BE

### PR DESCRIPTION
**Pull request by chandrasekharan-zipstack**
## What

- BE changes to support downloading a postman collection for API deployments
- FE changes to download postman collection
- MINOR: Refactored some exceptions in API deployments
- MINOR: Updated `HTTPMethod` enum to inherit `str`

NOTE: Unit tests can come in as a separate PR

## Why

- Helps easily download and import the collection in Postman to test the deployed APIs
- `HTTPMethod` had to be updated to support JSON serialization

## How

- Used dataclasses to model the Postman collection - relied on `WEB_APP_URL` setting to obtain the full path to the deployed API
- Set `Content-Disposition` header in the response to support `attachment` and passed the file name from the BE
- Added another option right after `Manage Keys` to download the collection

## Generated Postman Collection
[prompt studio test 2.json](https://github.com/Zipstack/unstract/files/15009498/prompt.studio.test.2.json)

## Relevant Docs

- [Postman collection v2 Format](https://learning.postman.com/collection-format/getting-started/overview/)
- [JSON serialization for enums](https://stackoverflow.com/a/51976841/12864260)
- [Good read on difference between .get() and .first()](https://www.helmut.dev/understanding-the-difference-between-get-and-first-in-django.html)
- [SO answer to help download the attachment from the FE with axios](https://stackoverflow.com/a/53230807/12864260)

## Notes on Testing

- Able to download the postman collection on API invocation, on importing the JSON in postman I see the collection and I'm able to execute the APIs right off the bat

## Screenshots
image
image
image
image
- Downloading from the FE
image
- Default name used in case the file name is not passed for some reason
image

NOTE: Toast messages are repeated due to React's strict mode, that can be ignored

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: backend/api/api_deployment_views.py [hunk 1]
```diff
@@ -119,9 +125,27 @@ def create(
             headers=headers,
         )
 
-
-def get_error_from_serializer(error_details: dict[str, Any]) -> Optional[str]:
-    error_key = next(iter(error_details))
-    # Get the first error message
-    error_message: str = f"{error_details[error_key][0]} : {error_key}"
-    return error_message
```
Comment (ritwik-g) [inline backend/api/api_deployment_views.py:127]: @chandrasekharan-zipstack why was this removed? @muhammad-ali-e is it safe do this?

File: backend/api/api_deployment_views.py [hunk 2]
```diff
@@ -119,9 +125,27 @@ def create(
...
+    @action(detail=True, methods=["get"])
+    def download_postman_collection(
+        self, request: Request, pk: Optional[str] = None
+    ) -> Response:
+        """Custom action to fetch a single instance."""
```
Comment (ritwik-g) [inline backend/api/api_deployment_views.py:132]: @chandrasekharan-zipstack is this doc string relevant to the method? Can you add a proper description?

File: backend/api/constants.py [hunk 3]
```diff
@@ -1,3 +1,6 @@
 class ApiExecution:
     PATH: str = "deployment/api"
     MAXIMUM_TIMEOUT_IN_SEC: int = 300  # 5 minutes
+    DEFAULT_TIMEOUT_IN_SEC: int = 80
+    FILES_FORM_DATA: str = "files"
```
Comment (ritwik-g) [inline backend/api/constants.py:5]: @muhammad-ali-e @chandrasekharan-zipstack should we use this constant here https://github.com/Zipstack/unstract/blob/779b944aab398b707cf71dd12d37058a2f5d1bdc/backend/api/api_deployment_views.py#L42C44-L42C49

File: backend/api/constants.py [hunk 4]
```diff
@@ -1,3 +1,6 @@
...
+    TIMEOUT_FORM_DATA: str = "timeout"
```
Comment (ritwik-g) [inline backend/api/constants.py:6]: @muhammad-ali-e @chandrasekharan-zipstack should we use this constant here https://github.com/Zipstack/unstract/blob/779b944aab398b707cf71dd12d37058a2f5d1bdc/backend/api/serializers.py#L71

File: backend/api/exceptions.py [hunk 5]
```diff
@@ -12,71 +12,44 @@ class ApiKeyCreateException(APIException):
     status_code = 500
     default_detail = "Exception while create API key"
 
-    def __init__(self, detail: Optional[str] = None, code: Optional[int] = None):
-        if detail is not None:
-            self.detail = detail
-        if code is not None:
-            self.code = code
-        super().__init__(detail, code)
-
```
Comment (ritwik-g) [inline backend/api/exceptions.py:21]: @chandrasekharan-zipstack why have we removed this init? Is it because the parent class have same implementation?

File: backend/api/constants.py [hunk 6]
```diff
@@ -1,3 +1,6 @@
 class ApiExecution:
     PATH: str = "deployment/api"
     MAXIMUM_TIMEOUT_IN_SEC: int = 300  # 5 minutes
+    DEFAULT_TIMEOUT_IN_SEC: int = 80
```
Comment (ritwik-g) [inline backend/api/constants.py:4]: @muhammad-ali-e @chandrasekharan-zipstack should we keep the default timeout to maximum ?

**Comment by ritwik-g** [comment]
@chandrasekharan-zipstack overall PR looks okay. Please check some of my queries. Also could you add the collection json generated from API in the PR description?

File: backend/api/api_deployment_views.py (hunk 1 above)
Comment (chandrasekharan-zipstack) [reply backend/api/api_deployment_views.py:127]: @ritwik-g we were previously using this to handle errors from serializers. This was incorrect and anyway we have replaced it with [drf-standardized-error](https://drf-standardized-errors.readthedocs.io/en/latest/index.html). So what I've removed is dead code - it wasn't being used anywhere either

File: backend/api/api_deployment_views.py [hunk 7]
```diff
@@ -119,9 +125,27 @@ def create(
...
+        instance = self.get_object()
+        active_api_key = ""
+        try:
+            api_key_inst = instance.apikey_set.filter(is_active=True).first()
```
Comment (muhammad-ali-e) [inline backend/api/api_deployment_views.py:136]: what is the apikey_set here? Is it the list of api keys of that API? can we use .get here?

File: backend/api/constants.py (hunk 3 above)
Comment (chandrasekharan-zipstack) [reply backend/api/constants.py:5]: Good catch, I wanted to change it but forgot. Will update it

File: backend/api/api_deployment_views.py [hunk 8]
```diff
@@ -119,9 +125,27 @@ def create(
...
+            active_api_key = api_key_inst.api_key
+        except APIKey.DoesNotExist:
```
Comment (muhammad-ali-e) [inline backend/api/api_deployment_views.py:138]: is .first() returns DoesNotExist error if it not exist? I think it return None or empty array.

File: backend/api/exceptions.py (hunk 5 above)
Comment (chandrasekharan-zipstack) [reply backend/api/exceptions.py:21]: Yes, this handling is redundant since DRF's [APIException](https://github.com/encode/django-rest-framework/blob/master/rest_framework/exceptions.py#L99) does it

File: backend/api/api_deployment_views.py [hunk 9]
```diff
@@ -119,9 +125,27 @@ def create(
...
+            logger.error(f"No active API key set for deployment {instance.pk}")
+            raise NoActiveAPIKeyError(deployment_name=instance.display_name)
+
+        postman_collection = PostmanCollection.create(
+            instance=instance, api_key=active_api_key
+        )
```
Comment (muhammad-ali-e) [inline backend/api/api_deployment_views.py:144]: NIT: I think It's better to do this functionality in the postman_collection directory. within some service or helper class

File: backend/api/constants.py (hunk 6 above)
Comment (chandrasekharan-zipstack) [reply backend/api/constants.py:4]: It can be done - however I remember that if the workflow runs within the timeout the API is synchronous and if its long-running we'll treat it as asynchronous and force the user to use the status API.
In this regard will it be a product concern? Anyway I feel that it can be a product call

File: backend/api/postman_collection/models.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
+from dataclasses import asdict, dataclass, field
```
Comment (muhammad-ali-e) [inline backend/api/postman_collection/models.py:1]: I think it's better to change this file name to dto.py or some other. I confuced this model.py with django db model.

File: backend/api/postman_collection/models.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+class PostmanItem:
+    name: str
+    request: RequestItem
+
+
+@dataclass
+class PostmanInfo:
+    name: str = "Unstract's API deployment"
+    schema: str = (
+        "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"  # noqa: E501
```
Comment (muhammad-ali-e) [inline backend/api/postman_collection/models.py:60]: What is this version v2.1.0? If it is avariable or want;s to change later, Make it a constant will be better,

File: backend/api/constants.py (hunk 6 above)
Comment (muhammad-ali-e) [reply backend/api/constants.py:4]: Also, How can we inform the maximum timeout value to user?

File: backend/api/api_deployment_views.py (hunk 8 above)
Comment (chandrasekharan-zipstack) [reply backend/api/api_deployment_views.py:138]: Yes you're right, responded on my action [here](https://github.com/Zipstack/unstract/pull/261#discussion_r1568376677)

File: backend/api/api_deployment_views.py (hunk 9 above)
Comment (chandrasekharan-zipstack) [reply backend/api/api_deployment_views.py:144]: I considered that briefly, however having the instance creation within the class is more pythonic - I've noticed libraries follow this and it also encapsulates the functions better rather than having a standalone helper outside.
I'd like us to break away from that practice wherever possible

File: backend/api/postman_collection/models.py (hunk 11 above)
Comment (chandrasekharan-zipstack) [reply backend/api/postman_collection/models.py:60]: This refers to the postman collection version that we're following. I'd rather make the entire string a constant

**Comment by chandrasekharan-zipstack** [comment]
@ritwik-g I've updated PR description with the generated postman collection

File: backend/api/constants.py (hunk 6 above)
Comment (ritwik-g) [reply backend/api/constants.py:4]: I think we have validations for the same. So we should be good I think

File: frontend/src/components/deployments/api-deployment/ApiDeployment.jsx [hunk 12]
```diff
@@ -275,6 +308,23 @@ function ApiDeployment() {
     },
     {
       key: "3",
+      label: (
+        <Space
+          direction="horizontal"
+          className="action-items"
+          onClick={() => downloadPostmanCollection()}
```
Comment (tahierhussain) [inline frontend/src/components/deployments/api-deployment/ApiDeployment.jsx:315]: @chandrasekharan-zipstack The usage of arrow function can be removed here. Can you please make the following change?

`onClick={downloadPostmanCollection}`

**Review by ritwik-g — APPROVED** [review]

File: frontend/src/components/deployments/api-deployment/ApiDeployment.jsx [hunk 13]
```diff
@@ -233,6 +234,38 @@ function ApiDeployment() {
       });
   };
 
+  const downloadPostmanCollection = () => {
+    apiDeploymentsApiService
+      .downloadPostmanCollection(selectedRow?.id)
+      .then((res) => {
+        const href = URL.createObjectURL(res?.data);
```
Comment (tahierhussain) [inline frontend/src/components/deployments/api-deployment/ApiDeployment.jsx:241]: @chandrasekharan-zipstack Suggestion: For better code quality, we can destructure the response and use the extracted variables wherever necessary.

`const { data, headers } = res;`

File: backend/api/constants.py (hunk 6 above)
Comment (chandrasekharan-zipstack) [reply backend/api/constants.py:4]: Anyway @ritwik-g do you want me to set the default timeout to the max possible?

**Review by tahierhussain — APPROVED** [review]
LGTM
